alf/algorithms/mdq_algorithm.py

```diff
@@ -0,0 +1,393 @@
+        alpha = self._log_alpha.exp().detach()
+
+        critic_input = (exp.observation, exp.action.to(torch.float32))
+
+        action = action.to(torch.float32)
+        target_critic_input = (exp.observation, action.detach())
+
+        critic, critic_state = self._critic_networks(
+            torch.cat(critic_input, -1),
+            alpha=alpha,
+            state=state.critic,
+            free_form=True)
+
+        noisy_distill_action = self._get_noisy_action(
+            action,
+            self._action_spec,
+            self._distill_noise,
+            noise_clip=0,
+            spec_clip=True)
+
+        critic_distill_input = (exp.observation, noisy_distill_action.detach())
+
+        critic_adv_form, critic_state = self._critic_networks(
+            critic_distill_input,
+            alpha=alpha,
+            state=state.critic,
+            free_form=False)
+        # [B, n, action_dim]
+        critic_adv_form = critic_adv_form.squeeze(3).permute(1, 2, 0)
+
+        # [B, n]
+        critic = critic.squeeze(2)
+
+        target_critic_input_new = (tensor_utils.tensor_extend_new_dim(
+            target_critic_input[0], dim=1, n=self._num_critic_replicas),
+                                   target_critic_input[1])
+
+        distill_critic_input_new = (tensor_utils.tensor_extend_new_dim(
+            critic_distill_input[0], dim=1, n=self._num_critic_replicas),
+                                    critic_distill_input[1])
+
+        target_critic, target_critic_state = self._target_critic_networks(
+            torch.cat(target_critic_input_new, -1),
+            alpha=alpha,
+            state=state.target_critic,
+            free_form=True)
+
+        distill_target, _ = self._target_critic_networks(
+            torch.cat(distill_critic_input_new, -1),
+            alpha=alpha,
+            state=state.target_critic,
+            free_form=True)
+
+        distill_target = distill_target.reshape(distill_target.shape[0], -1, 1)
+
+        target_critic = target_critic.squeeze(2)
+        distill_target = distill_target.squeeze(2)
+
+        kl_wrt_prior_i = log_pi - self._log_pi_uniform_prior
+        # keeping the KL of all actions dimensions in case it is useful
+        # in some cases in the future, e.g., per-action target correction using
+        # the corresponding KL
+        kl_wrt_prior = tensor_utils.reverse_cumsum(kl_wrt_prior_i, dim=-1)
+        info = MdqCriticInfo(
+            critic_free_form=critic,
+            target_critic_free_form=target_critic,
+            distill_target=distill_target,
+            critic_adv_form=critic_adv_form,
+            kl_wrt_prior=kl_wrt_prior)
+
+        state = MdqCriticState(
+            critic=critic_state, target_critic=target_critic_state)
+
+        return state, info
+
+    def _alpha_train_step(self, log_pi):
+        log_pi_cum = tensor_utils.reverse_cumsum(log_pi, dim=-1)
+        log_pi_full = log_pi_cum[..., 0:1]
```

Comment:
since log_pi_cum is not used elsewhere, log_pi_full can simply be `log_pi.sum(dim=-1, keepdim=True)`

Reply:
cumsum is unnecessary. Replaced with sum.